rust/datafusion/src/physical_plan/parquet.rs

```diff
@@ -30,62 +30,117 @@ use crate::physical_plan::{common, Partitioning};
+                "No Parquet files found at path {}",
+                path
+            )))
         } else {
-            let file = File::open(&filenames[0])?;
-            let file_reader = Arc::new(SerializedFileReader::new(file)?);
-            let mut arrow_reader = ParquetFileArrowReader::new(file_reader);
-            let schema = arrow_reader.get_schema()?;
-            let metadata = arrow_reader.get_metadata();
-
-            Ok(Self::new(
-                filenames, schema, projection, batch_size, metadata,
-            ))
+            // build a list of Parquet partitions with statistics and gather all unique schemas
+            // used in this data set
+            let mut schemas: Vec<Schema> = vec![];
+            let mut partitions = vec![];
+            for filename in &filenames {
+                let file = File::open(filename)?;
+                let file_reader = Arc::new(SerializedFileReader::new(file)?);
+                let mut arrow_reader = ParquetFileArrowReader::new(file_reader);
+                let meta_data = arrow_reader.get_metadata();
+                // collect all the unique schemas in this data set
+                let schema = arrow_reader.get_schema()?;
+                if schemas.is_empty() || schema != schemas[0] {
+                    schemas.push(schema);
+                }
+                let mut num_rows = 0;
+                let mut total_byte_size = 0;
+                for i in 0..meta_data.num_row_groups() {
+                    let row_group_meta = meta_data.row_group(i);
+                    num_rows += row_group_meta.num_rows();
+                    total_byte_size += row_group_meta.total_byte_size();
+                }
+                let statistics = Statistics {
+                    num_rows: Some(num_rows as usize),
+                    total_byte_size: Some(total_byte_size as usize),
+                };
+                partitions.push(ParquetPartition {
+                    filename: filename.to_owned(),
+                    statistics,
+                });
+            }
+
+            // we currently get the schema information from the first file rather than do
+            // schema merging and this is a limitation.
+            // See https://issues.apache.org/jira/browse/ARROW-11017
+            if schemas.len() > 1 {
+                return Err(DataFusionError::Plan(format!(
+                    "The Parquet files in {} have {} different schemas and DataFusion does \
+                    not yet support schema merging", path, schemas.len())));
+            }
+            let schema = schemas[0].clone();
+
+            Ok(Self::new(partitions, schema, projection, batch_size))
         }
     }
 
-    /// Create a new Parquet reader execution plan with provided files and schema
+    /// Create a new Parquet reader execution plan with provided partitions and schema
     pub fn new(
-        filenames: Vec<String>,
+        partitions: Vec<ParquetPartition>,
```

Comment:
I thought that part of the reason for allowing a list of files was to support similar behavior to DeltaLake where an external file contains a list of filenames representing a version of data rather than grouping them by directory structure?

Reply:
Interesting. I was not aware of this use case. Perhaps we need a specific constructor for that use case. I'll take a look.

Reply:
It may be premature to support that use case anyway until more of the core engine works.

Reply:
I pushed a change so there are now `try_from_path` and `try_from_files` constructors

Reply:
Looks good 👍